core/src/main/java/org/elasticsearch/node/AdaptiveSelectionStats.java

```diff
@@ -0,0 +1,111 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.node;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.util.set.Sets;
+import org.elasticsearch.common.xcontent.ToXContent.Params;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Class representing statistics about adaptive replica selection. This includes
+ * EWMA of queue size, service time, and response time, as well as outgoing
+ * searches to each node and the "rank" based on the ARS formula.
+ */
+public class AdaptiveSelectionStats implements Writeable, ToXContentFragment {
+
+    private final Map<String, Long> clientOutgoingConnections;
+    private final Map<String, ResponseCollectorService.ComputedNodeStats> nodeComputedStats;
+
+    public AdaptiveSelectionStats(Map<String, Long> clientConnections,
+                                  Map<String, ResponseCollectorService.ComputedNodeStats> nodeComputedStats) {
+        this.clientOutgoingConnections = clientConnections;
+        this.nodeComputedStats = nodeComputedStats;
+    }
+
+    public AdaptiveSelectionStats(StreamInput in) throws IOException {
+        this.clientOutgoingConnections = in.readMap(StreamInput::readString, StreamInput::readLong);
+        this.nodeComputedStats = in.readMap(StreamInput::readString, ResponseCollectorService.ComputedNodeStats::new);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeMap(this.clientOutgoingConnections, StreamOutput::writeString, StreamOutput::writeLong);
+        out.writeMap(this.nodeComputedStats, StreamOutput::writeString, (stream, stats) -> stats.writeTo(stream));
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject("adaptive_selection");
+        Set<String> allNodeIds = Sets.union(clientOutgoingConnections.keySet(), nodeComputedStats.keySet());
+        for (String nodeId : allNodeIds) {
+            builder.startObject(nodeId);
+            if (clientOutgoingConnections.containsKey(nodeId)) {
+            }
+            ResponseCollectorService.ComputedNodeStats stats = nodeComputedStats.get(nodeId);
```

Comment:
This empty `if` followed by this line looks off.

Reply:
Whoops yeah, that's a leftover, I'll remove it